File: apps/extractor/app/overrides.py
#!/usr/bin/env python3
"""
The override layer: what a human corrects after the pipeline has decided.

Inputs are what the system does not know. Overrides are what it *decided* — and
on the reference budget it decides a great deal without being asked: 420 phase
lines whose timing is a guess, 218 inferred rate bases, 31 multipliers rebuilt
arithmetically. Every one of those needs an escape hatch.

Three rules hold this together.

  1. **Reconciliation is sacred.** The whole trust argument is "this sums to the
     budget". So a `redistribute` override may change *when* money lands and
     never how much, and the total is asserted afterwards. Changing an amount is
     an `amend`, which is versioned, carries a reason, and is reported as a
     departure from the budget rather than hidden inside it.

  2. **Key on identity, never on position.** Overrides have to survive re-import
     because the product is the weekly reforecast. Budgets re-version underneath
     you — 30 of 35 departments moved between two versions of this production
     while the total held — so row numbers are useless as keys.

  3. **A correction is not an override.** "The parser got it wrong" is a bug
     report; "I know something no document states" is permanent knowledge. They
     look identical at the moment of editing and diverge completely over time.

Usage:
    from overrides import OverrideSet
    overrides = OverrideSet.load("overrides.json")
    value = overrides.resolve("timing", target, default=computed)
    overrides.report()
"""

from __future__ import annotations

import json
from collections import defaultdict
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any

# Most specific first — the order in which a target's keys are tried.
SCOPES = ("line", "account", "department", "production", "company")

REDISTRIBUTE = "redistribute"   # changes timing; total preserved
AMEND = "amend"                 # changes an amount; total moves

CORRECTION = "correction"       # the parser was wrong — a bug with a repro
JUDGEMENT = "judgement"         # the human knows what no document states


@dataclass
class Override:
    field: str                        # timing | phase | rate_basis | day_cost | ...
    value: Any
    scope: str                        # one of SCOPES
    key: str                          # the identity this applies to
    kind: str = REDISTRIBUTE          # REDISTRIBUTE | AMEND
    origin: str = JUDGEMENT           # CORRECTION | JUDGEMENT
    reason: str | None = None
    author: str | None = None
    created: str | None = None        # ISO date, supplied by the caller

    def matches(self, field: str, keys: dict[str, str]) -> bool:
        return self.field == field and keys.get(self.scope) == self.key


def target_keys(*, department: str | None = None, account: str | None = None,
                sub: str | None = None, person: str | None = None,
                production: str | None = None,
                company: str | None = None) -> dict[str, str]:
    """Identity for a thing an override can attach to, most specific first.

    Degrades deliberately: a line key falls back to its account, then its
    department. A user who says "Camera preps two weeks earlier" should not have
    to repeat themselves for every person in Camera.
    """
    keys: dict[str, str] = {}
    if account and (sub or person):
        keys["line"] = f"{account}|{sub or ''}|{(person or '').strip().upper()}"
    if account:
        keys["account"] = account
    if department:
        keys["department"] = department
    if production:
        keys["production"] = production
    if company:
        keys["company"] = company
    return keys


@dataclass
class OverrideSet:
    overrides: list[Override] = field(default_factory=list)
    applied: dict[str, int] = field(default_factory=lambda: defaultdict(int))
    unused: list[Override] = field(default_factory=list)
    amendments: list[Override] = field(default_factory=list)

    # ---------- loading ------------------------------------------------------

    @classmethod
    def load(cls, path: str | Path | None) -> "OverrideSet":
        if not path or not Path(path).exists():
            return cls()
        payload = json.loads(Path(path).read_text())
        entries = [Override(**{k: v for k, v in item.items()
                               if k in Override.__dataclass_fields__})
                   for item in payload.get("overrides", [])]
        bad = [o for o in entries if o.scope not in SCOPES]
        if bad:
            raise SystemExit(
                f"unknown scope(s): {sorted({o.scope for o in bad})}; "
                f"expected one of {SCOPES}")
        return cls(overrides=entries)

    def save(self, path: str | Path) -> None:
        Path(path).write_text(json.dumps(
            {"overrides": [asdict(o) for o in self.overrides]}, indent=2))

    # ---------- resolution ---------------------------------------------------

    def resolve(self, field_name: str, keys: dict[str, str], default: Any = None,
                *, allow_amend: bool = False) -> Any:
        """Most specific override wins; otherwise the computed default stands."""
        for scope in SCOPES:
            if scope not in keys:
                continue
            for override in self.overrides:
                if override.scope != scope or not override.matches(field_name, keys):
                    continue
                if override.kind == AMEND and not allow_amend:
                    # An amendment changes an amount, so it cannot be applied
                    # silently inside a redistribution pass.
                    self.amendments.append(override)
                    continue
                self.applied[f"{field_name}@{scope}"] += 1
                return override.value
        return default

    # ---------- integrity ----------------------------------------------------

    def collect_amendments(self) -> list[Override]:
        """Surface every amendment, whether or not a resolver asked for it.

        An amendment changes an amount, so it cannot be applied inside a
        redistribution pass — but it must never be silently dropped either. If
        nothing in the pipeline consults that field, the user has still asked
        for a change and is owed an answer.
        """
        seen = {id(o) for o in self.amendments}
        for override in self.overrides:
            if override.kind == AMEND and id(override) not in seen:
                self.amendments.append(override)
        return self.amendments

    def find_orphans(self, live_keys: set[str]) -> list[Override]:
        """Overrides pointing at something the current budget no longer has.

        Dropping these silently produces a wrong number nobody can see; keeping
        them silently is worse. They are surfaced and left for a human.
        """
        self.unused = [o for o in self.overrides if o.key not in live_keys]
        return self.unused

    @staticmethod
    def assert_total_preserved(before: float, after: float,
                               tolerance: float = 1.0) -> None:
        """Redistribution must not move the total. Fail loudly if it did."""
        if abs(before - after) > tolerance:
            raise SystemExit(
                f"redistribution changed the total by {after - before:,.2f} "
                f"({before:,.2f} -> {after:,.2f}). A redistribute override may "
                f"change when money lands, never how much. Use kind='amend' if "
                f"the amount really is changing.")

    # ---------- reporting ----------------------------------------------------

    def report(self) -> None:
        if not self.overrides:
            print("\nno overrides loaded")
            return
        corrections = sum(1 for o in self.overrides if o.origin == CORRECTION)
        judgements = len(self.overrides) - corrections
        print(f"\noverrides: {len(self.overrides)} loaded · "
              f"{sum(self.applied.values())} applied")
        print(f"   {judgements} judgement (permanent), "
              f"{corrections} correction (parser bug — should trend to zero)")
        for key, count in sorted(self.applied.items(), key=lambda x: -x[1]):
            print(f"   {key:<34}{count:>4}")
        if self.amendments:
            print(f"\n{len(self.amendments)} AMENDMENT(S) — these change amounts, "
                  f"so they are reported and NOT applied. Each needs a versioned "
                  f"budget amendment before it can affect the schedule:")
            for o in self.amendments[:6]:
                print(f"   {o.scope}:{o.key} {o.field} → {o.value}"
                      f"   {o.reason or 'no reason given'}")
        if self.unused:
            print(f"\n{len(self.unused)} ORPHANED — the budget no longer has "
                  f"what these point at:")
            for o in self.unused[:6]:
                print(f"   {o.scope}:{o.key} {o.field}"
                      f"   ({o.author or 'unknown'}, {o.created or 'undated'})")


def example() -> dict[str, Any]:
    """A worked example, which is also the file format."""
    return {
        "overrides": [
            {"field": "prep_lead_weeks", "value": 2, "scope": "department",
             "key": "3300", "kind": REDISTRIBUTE, "origin": JUDGEMENT,
             "reason": "Camera preps two weeks ahead on this show",
             "author": "line producer", "created": "2017-09-05"},
            {"field": "phase_window", "value": "prep", "scope": "account",
             "key": "2519", "kind": REDISTRIBUTE, "origin": JUDGEMENT,
             "reason": "crane wanted a week early for rigging",
             "author": "key grip", "created": "2017-09-12"},
            {"field": "rate_basis", "value": "hour", "scope": "line",
             "key": "3302|*002/\"B\"CAMERAOP|CRAIG BAUER", "kind": REDISTRIBUTE,
             "origin": CORRECTION,
             "reason": "parser read the loaded weekly figure as the rate",
             "author": "accountant", "created": "2017-09-18"},
            {"field": "amount", "value": 42000, "scope": "account",
             "key": "2400", "kind": AMEND, "origin": JUDGEMENT,
             "reason": "set strike re-quoted after the location change",
             "author": "UPM", "created": "2017-10-02"},
        ]
    }


if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1 and sys.argv[1] == "--example":
        print(json.dumps(example(), indent=2))
    else:
        print(__doc__)

File: apps/api/src/routes/overrides.js
/**
 * Adjustments — what a person corrected after the generator decided.
 *
 * The engine has always taken these; nothing in the interface ever sent one, so
 * "I know this rig comes in a week early and the schedule does not" had no way
 * of being said. They are stored per production rather than posted with each
 * generation, because the product is the weekly reforecast and a judgement made
 * in prep should still hold in week six.
 *
 * Only redistributions can be made here. An amendment changes how much money
 * there is, which belongs in a versioned budget rather than in a schedule, and
 * the generator reports rather than applies them.
 */

import { Router } from 'express';
import { prisma } from '../db.js';
import { requireRole } from '../auth/middleware.js';

export const overridesRouter = Router({ mergeParams: true });

// The fields the generator actually consults. Offering any others would be
// inventing a control that silently does nothing.
export const FIELDS = {
  phase_window: {
    label: 'Move it to a different phase',
    help: 'Use this for equipment. Most rentals are budgeted as an allowance over '
      + 'a stated hire, which the schedule places against the shoot. Moving one to '
      + 'prep is how a crane wanted a week early for rigging lands a week early.',
    type: 'choice',
    choices: ['prep', 'shoot', 'wrap', 'post'],
    scopes: ['department', 'account'],
  },
  prep_lead_weeks: {
    label: 'Start prep earlier',
    help: 'Pushes prep spend further ahead of the shoot. This only moves money '
      + 'the budget already treats as prep — for a rental billed as an allowance, '
      + 'move it to prep first and then set this if it needs to be earlier still.',
    type: 'number',
    unit: 'weeks',
    scopes: ['department', 'account'],
  },
  cash_class: {
    label: 'Settles on different terms',
    help: 'Which payment lag this money follows. Prepaid leaves before the work, '
      + 'labour follows payroll, vendor follows invoice terms.',
    type: 'choice',
    choices: ['prepaid', 'labour', 'vendor'],
    scopes: ['department', 'account'],
  },
};

const SCOPES = ['line', 'account', 'department', 'production', 'company'];

function publicOverride(o) {
  return {
    id: o.id,
    field: o.field,
    value: o.value,
    scope: o.scope,
    key: o.key,
    kind: o.kind,
    origin: o.origin,
    reason: o.reason,
    author: o.authorName,
    createdAt: o.createdAt,
  };
}

/** The shape the extractor's OverrideSet.load expects. */
export async function overridePayloadFor(productionId) {
  const rows = await prisma.productionOverride.findMany({
    where: { productionId },
    orderBy: { createdAt: 'asc' },
  });
  if (!rows.length) return null;
  return {
    overrides: rows.map((o) => ({
      field: o.field,
      value: o.value,
      scope: o.scope,
      key: o.key,
      kind: o.kind,
      origin: o.origin,
      reason: o.reason,
      author: o.authorName,
      created: o.createdAt.toISOString().slice(0, 10),
    })),
  };
}

/** What can be adjusted, and what can be adjusted about it. */
overridesRouter.get('/fields', (_req, res) => {
  res.json({ fields: FIELDS });
});

overridesRouter.get('/', async (req, res, next) => {
  try {
    const rows = await prisma.productionOverride.findMany({
      where: { productionId: req.production.id },
      orderBy: { createdAt: 'desc' },
    });
    res.json(rows.map(publicOverride));
  } catch (error) {
    next(error);
  }
});

overridesRouter.post('/', requireRole('editor'), async (req, res, next) => {
  try {
    const field = String(req.body?.field || '');
    const scope = String(req.body?.scope || '');
    const key = String(req.body?.key || '').trim();
    const reason = String(req.body?.reason || '').trim();
    const origin = req.body?.origin === 'correction' ? 'correction' : 'judgement';
    let value = req.body?.value;

    const spec = FIELDS[field];
    if (!spec) {
      return res.status(400).json({
        error: `Unknown adjustment. This budget understands: ${Object.keys(FIELDS).join(', ')}.`,
      });
    }
    if (!SCOPES.includes(scope) || !spec.scopes.includes(scope)) {
      return res.status(400).json({
        error: `${spec.label} applies to ${spec.scopes.join(' or ')}, not ${scope || 'nothing'}.`,
      });
    }
    if (!key) return res.status(400).json({ error: 'Choose what this applies to.' });
    // A reason is the point. An adjustment nobody can explain in six weeks is
    // indistinguishable from a mistake.
    if (reason.length < 3) {
      return res.status(400).json({ error: 'Say why — this is read months later.' });
    }
    if (spec.type === 'number') {
      value = Number(value);
      if (!Number.isFinite(value) || value < 0 || value > 52) {
        return res.status(400).json({ error: 'Give a number of weeks between 0 and 52.' });
      }
    } else if (!spec.choices.includes(String(value))) {
      return res.status(400).json({
        error: `${spec.label} must be one of: ${spec.choices.join(', ')}.`,
      });
    }

    const saved = await prisma.productionOverride.upsert({
      where: {
        productionId_field_scope_key: {
          productionId: req.production.id, field, scope, key,
        },
      },
      update: { value, reason, origin, authorId: req.user.id, authorName: req.user.name || req.user.email },
      create: {
        productionId: req.production.id,
        field,
        value,
        scope,
        key,
        // Nothing here can change a total; that is the whole trust argument.
        kind: 'redistribute',
        origin,
        reason,
        authorId: req.user.id,
        authorName: req.user.name || req.user.email,
      },
    });
    return res.status(201).json(publicOverride(saved));
  } catch (error) {
    return next(error);
  }
});

overridesRouter.delete('/:overrideId', requireRole('editor'), async (req, res, next) => {
  try {
    const row = await prisma.productionOverride.findUnique({
      where: { id: req.params.overrideId },
    });
    if (!row || row.productionId !== req.production.id) {
      return res.status(404).json({ error: 'Adjustment not found.' });
    }
    await prisma.productionOverride.delete({ where: { id: row.id } });
    return res.json({ ok: true });
  } catch (error) {
    return next(error);
  }
});
